### PR TITLE
Refactor the setting of tbuffer format

### PR DESCRIPTION
**Pull request by xuechen417**
Remove the duplicate code of setting tbuffer format for creating
"llvm.amdgcn.raw.tbuffer.*" each time. Use an ArrayRef `m_buffFormat`
referring to one of two static const arrays (one for <= gfx9, one for >=
gfx10) `BufferFormats[2][4]`.

### CODE REVIEW

File: lgc/patch/PatchInOutImportExport.cpp [hunk 1]
```diff
@@ -3991,7 +3975,7 @@ void PatchInOutImportExport::storeValueToGsVsRing(Value *storeValue, unsigned lo
             ConstantInt::get(Type::getInt32Ty(*m_context), coherent.u32All) // glc, slc, swz
         };
         emitCall("llvm.amdgcn.raw.tbuffer.store.i32", Type::getVoidTy(*m_context), args, {}, insertPos);
-      } else if (m_gfxIp.major == 10) {
+      } else {
```
Comment (jayfoad) [inline lgc/patch/PatchInOutImportExport.cpp:3978]: This looks like an unrelated change.

**Comment by jayfoad** [comment]
Looks OK to me. Perhaps m_buffFormats could even be a `std::array<unsigned, 4> *` so the length of the array is a known compile-time constant. But it doesn't really matter.

File: lgc/patch/PatchInOutImportExport.cpp (hunk 1 above)
Comment (xuechen417) [reply lgc/patch/PatchInOutImportExport.cpp:3978]: The original code is as below:
```
else if (m_gfxip.major == 10) else {
        CoherentFlag coherent = {};
        coherent.bits.glc = true;
        coherent.bits.slc = true;
        coherent.bits.swz = true;
        Value *args[] = {
            storeValue,                                                          // vdata
            m_pipelineSysValues.get(m_entryPoint)->getGsVsRingBufDesc(streamId), // rsrc
            ringOffset,                                                          // voffset
            gsVsOffset,                                                          // soffset
            ConstantInt::get(Type::getInt32Ty(*m_context), BUF_FORMAT_32_UINT),  // format
            ConstantInt::get(Type::getInt32Ty(*m_context), coherent.u32All)      // glc, slc, swz
        }; 
```
I think the combined buffer format is used for Gfx10+, is right?

**Review by xuechen417 — COMMENTED** [review]
V1: Change the data type to std::array<unsigned, 4> * for m_buffFormat.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 2cbe948c56db97b47061dc3b259746aeafffccad
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 8abed125acda77ccf700ddd0f5bf0114d25ac135</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 2f28502c5e6e4f059a014ab30001a191328f7a5d</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 9ef380b0ac6d8e8aa7351981bce99a8a173d3632</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/224557)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 26814/56120 (47.8%)</li>
<li>Failed: 0/56120 (0.0%)</li>
<li>Skipped: 29306/56120 (52.2%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26814/56120 (47.8%)</li>
<li>Failed: 0/56120 (0.0%)</li>
<li>Skipped: 29306/56120 (52.2%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 25403/56197 (45.2%)</li>
<li>Failed: 0/56197 (0.0%)</li>
<li>Skipped: 30794/56197 (54.8%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 26829/56120 (47.8%)</li>
<li>Failed: 0/56120 (0.0%)</li>
<li>Skipped: 29291/56120 (52.2%)</li>
</ul></details>


</ul></details>

File: lgc/patch/PatchInOutImportExport.cpp (hunk 1 above)
Comment (jayfoad) [reply lgc/patch/PatchInOutImportExport.cpp:3977]: Yes the combined buffer format is used for GFX11.

File: lgc/patch/PatchInOutImportExport.cpp [hunk 2]
```diff
@@ -113,6 +113,18 @@ bool PatchInOutImportExport::runOnModule(Module &module) {
   if (m_hasTs || (m_hasGs && (m_pipelineState->isGsOnChip() || m_gfxIp.major >= 9)))
     m_lds = Patch::getLdsVariable(m_pipelineState, m_module);
 
+  // Set buffer formats based on specific GFX
+  static const std::array<std::array<unsigned, 4>, 2> BufferFormats = {
+      {{
+           ((BUF_NUM_FORMAT_FLOAT << 4) | (BUF_DATA_FORMAT_32)),
```
Comment (jayfoad) [inline lgc/patch/PatchInOutImportExport.cpp:119]: You don't need any of these parentheses.

File: lgc/patch/PatchInOutImportExport.cpp [hunk 3]
```diff
@@ -113,6 +113,18 @@ bool PatchInOutImportExport::runOnModule(Module &module) {
...
+           ((BUF_NUM_FORMAT_FLOAT << 4) | (BUF_DATA_FORMAT_32_32)),
+           ((BUF_NUM_FORMAT_FLOAT << 4) | (BUF_DATA_FORMAT_32_32_32)),
+           ((BUF_NUM_FORMAT_FLOAT << 4) | (BUF_DATA_FORMAT_32_32_32_32)),
+       },
+       {BUF_FORMAT_32_FLOAT, BUF_FORMAT_32_32_FLOAT, BUF_FORMAT_32_32_32_FLOAT, BUF_FORMAT_32_32_32_32_FLOAT}}};
+
+  m_buffFormats = m_gfxIp.major <= 9 ? &(BufferFormats[0]) : &(BufferFormats[1]);
```
Comment (jayfoad) [inline lgc/patch/PatchInOutImportExport.cpp:126]: You don't need the parentheses.

**Review by jayfoad — APPROVED** [review]

File: lgc/patch/PatchInOutImportExport.cpp (hunk 2 above)
Comment (xuechen417) [reply lgc/patch/PatchInOutImportExport.cpp:119]: Fixed.

File: lgc/patch/PatchInOutImportExport.cpp (hunk 3 above)
Comment (xuechen417) [reply lgc/patch/PatchInOutImportExport.cpp:126]: Fixed.

**Review by xuechen417 — COMMENTED** [review]
V2: Remove redundant parentheses.

**Review by jayfoad — APPROVED** [review]

**Comment by JaxLinAMD** [comment]
@xuechen417 could you rebase your branch xuechen417:buffFormat onto latest dev and force push it to avoid build error?

**Comment by xuechen417** [comment]
V3: Rebase. Thanks @JaxLinAMD for your reminder.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 790595c2b28ea0828f3e524304d7e122586eaf73
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 1e61f9ab6eb2041c73d7e7bd8f2021fa05a4bdd8</li>
<li>SPVGEN: dev 7c1cf34dd7eb4d3a6e6652c26292eef1bfcbcaf0</li>
<li>XGL: dev 3237b046add2809634e7f5f3fcdea1af4ad6613d</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 294f18dfb21b21ca76f4b742bffa3177babe46ff</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/224557)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 26814/56120 (47.8%)</li>
<li>Failed: 0/56120 (0.0%)</li>
<li>Skipped: 29306/56120 (52.2%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26814/56120 (47.8%)</li>
<li>Failed: 0/56120 (0.0%)</li>
<li>Skipped: 29306/56120 (52.2%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 25403/56197 (45.2%)</li>
<li>Failed: 0/56197 (0.0%)</li>
<li>Skipped: 30794/56197 (54.8%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 26829/56120 (47.8%)</li>
<li>Failed: 0/56120 (0.0%)</li>
<li>Skipped: 29291/56120 (52.2%)</li>
</ul></details>


</ul></details>
